vm-migration/src/bitpos_iterator.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright © 2025 Cyberus Technology GmbH
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// An iterator that turns a sequence of u64s into a sequence of bit positions
+/// that are set.
+///
+/// This is useful to iterate over dirty memory bitmaps.
+struct BitposIterator<I> {
+    underlying_it: I,
+
+    /// How many u64's we've already consumed.
+    word_pos: usize,
+
+    /// If we already started working on a u64, it's here. Together with the bit
+    /// position where we have to continue.
+    current_word: Option<(u64, u32)>,
+}
+
+impl<I> Iterator for BitposIterator<I>
+where
+    I: Iterator<Item = u64>,
+{
+    type Item = u64;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if self.current_word.is_none() {
+                self.current_word = self.underlying_it.next().map(|w| (w, 0));
+            }
+
+            let (word, word_bit) = self.current_word?;
+
+            // Continue early if there is no chance to find something.
+            if word != 0 && word_bit < 64 {
+                let shifted_word = word >> word_bit;
+                if shifted_word != 0 {
+                    let zeroes = shifted_word.trailing_zeros();
+
+                    self.current_word = Some((word, zeroes + word_bit + 1));
+                    let next_bitpos =
+                        u64::try_from(self.word_pos).unwrap() * 64 + u64::from(word_bit + zeroes);
```

Comment:
I think `word_pos` can be defined as `u32` (4 billion bits is more than enough). Together with `<<`, you can make this operation safer (perhaps faster).

Also, I'd do a checked_add with `unwrap` to be safer too, which should never happen with our implementation here.